File: Mono/Tables/TableSource.cs
using System;

namespace Tables
{
    public class TableSource
    {
        private Object data;
        public TableRowType DefaultStringRowType;

        public TableSource(Object data=null,TableRowType defaultStringRowType=TableRowType.Text)
        {
            this.data = data;
            this.DefaultStringRowType = defaultStringRowType;
        }

        public Object Data
        {
            get
            {
                return data;
            }
        }

        public int RowsInSection(int section)
        {
            if (data != null)
            {
                return data.GetType().GetProperties().Length;
            }
            return 0;
        }

        public string GetName(int row, int section)
        {
            var prop = data.GetType().GetProperties()[row];
            return prop.Name;
        }

        public Object GetValue(int row, int section)
        {
            var prop = data.GetType().GetProperties()[row];
            return prop.GetValue(data);
        }

        public void SetValue(Object obj, int row, int section)
        {
            var prop = data.GetType().GetProperties()[row];
            prop.SetValue(data,obj);
        }

        public string DisplayName(ITableAdapterRowConfigurator configurator,int row,int section)
        {
            var prop = data.GetType().GetProperties()[row];
            var s = RowSetting(configurator,prop.Name);
            if (s != null && s.DisplayName != null)
                return s.DisplayName;
            return prop.Name;
        }

        public string DisplayDate(ITableAdapterRowConfigurator configurator, int row, int section, DateTime date, TableRowType rowType)
        {
            var prop = data.GetType().GetProperties()[row];
            var s = RowSetting(configurator,prop.Name);

            string format = null;

            switch (rowType)
            {
                case TableRowType.Date:
                    format = "d MMM yyyy";
                break;
                case TableRowType.Time:
                    format = "HH:mm";
                break;
                case TableRowType.DateTime:
                    format = "HH:mm d MMM yyyy";
                break;
            }

            if (s != null)
            {
                switch (rowType)
                {
                    case TableRowType.Date:
                        if (s.DateFormat!=null)
                            format = s.DateFormat;
                    break;
                    case TableRowType.Time:
                        if (s.TimeFormat!=null)
                            format = s.TimeFormat;
                    break;
                    case TableRowType.DateTime:
                        if (s.DateTimeFormat!=null)
                            format = s.DateTimeFormat;
                    break;
                }
            }
            return format == null ? date.ToString() : date.ToString(format);
        }

        public TableRowType RowType(ITableAdapterRowConfigurator configurator,int row,int section)
        {
            var prop = data.GetType().GetProperties()[row];
            var ptype = prop.PropertyType;
            var s = RowSetting(configurator,prop.Name);
            if (s != null && s.RowType != TableRowType.Unknown)
                return s.RowType;
            var rowType = DefaultStringRowType;
            if (ptype == typeof(string))
                rowType = DefaultStringRowType;
            else if (ptype == typeof(bool))
                rowType = TableRowType.Checkbox;
            else if (ptype == typeof(DateTime))
                rowType = TableRowType.DateTime;
            return rowType;
        }

        public TableAdapterRowConfig RowSetting(ITableAdapterRowConfigurator configurator,string rowName)
        {
            TableAdapterRowConfig settings = null;
            if (data is ITableAdapterRowConfigurator)
                settings = (data as ITableAdapterRowConfigurator).ConfigForRow(rowName);
            else if (configurator != null)
                settings = configurator.ConfigForRow(rowName);
            return settings;
        }

        public bool Editable(ITableAdapterRowConfigurator configurator,int row,int section)
        {
            var prop = data.GetType().GetProperties()[row];
            var name = prop.Name;
            var editable = prop.CanWrite && prop.SetMethod.IsPublic;
            TableAdapterRowConfig settings = RowSetting(configurator,name);
            if (settings!=null) 
                editable = settings.Editable;
            return editable;
        }
    }
}

File: objc/Tables/TableSource.m
//
//  TableSource.m
//  tables
//
//  Created by Aleksander Slater on 07/09/2014.
//  Copyright (c) 2014 Davincium. All rights reserved.
//

#import "TableSource.h"
#import "TablePropertyScanner.h"

@interface TableSource()
@property(nonatomic,strong) NSMutableDictionary*formatters;
@property(nonatomic,strong) NSMutableArray*properties;
@end

@implementation TableSource

- (id)initWithData:(NSObject*)data defaultStringRowType:(TableRowType)rowType
{
    self = [super init];
    if (self)
    {
        self.formatters = [[NSMutableDictionary alloc] init];
        self.properties = [[NSMutableArray alloc] init];
        self.data = data;
        self.defaultStringRowType= rowType;
        
        [self reloadProperties];
    }
    return self;
}

- (id)initWithData:(NSObject*)data
{
    self = [self initWithData:data defaultStringRowType:TableRowTypeText];
    if (self)
    {
        
    }
    return self;
}

- (void)reloadProperties
{
    [self.properties removeAllObjects];
    NSArray *objectProperties = [TablePropertyScanner propertiesForObject:self.data];
    for (NSDictionary*d in objectProperties)
    {
        [self.properties addObject:[d objectForKey:TablePropertyScannerKey]];
    }
}

- (NSDateFormatter*)formatterForFormat:(NSString*)format
{
    NSDateFormatter *f = [self.formatters objectForKey:format];
    if (f==nil)
    {
        f = [[NSDateFormatter alloc] init];
        [f setDateFormat:format];
        [self.formatters setObject:f forKey:format];
    }
    return f;
}

- (NSInteger)numberOfSections
{
    return 1;
}

- (NSInteger)rowsInSection:(NSInteger)section
{
    if (self.data != nil)
    {
        return self.properties.count;
    }
    return 0;
}

- (NSObject*)getValueAtRow:(NSInteger)row andSection:(NSInteger)section
{
    NSString *key = self.properties[row];
    NSObject *obj = [self.data valueForKey:key];
    return obj;
}

- (NSString*)getValueStringAtRow:(NSInteger)row andSection:(NSInteger)section
{
    NSObject *value = [self getValueAtRow:row andSection:section];
    NSString *text  = @"";
    if ([value isKindOfClass:[NSString class]])
        text = (NSString*)value;
    return text;
}

- (NSNumber*)getValueNumberAtRow:(NSInteger)row andSection:(NSInteger)section
{
    NSObject *value = [self getValueAtRow:row andSection:section];
    if ([value isKindOfClass:[NSNumber class]])
        return (NSNumber*)value;
    return nil;
}

- (NSDate*)getValueDateAtRow:(NSInteger)row andSection:(NSInteger)section
{
    NSObject *value = [self getValueAtRow:row andSection:section];
    if ([value isKindOfClass:[NSDate class]])
        return (NSDate*)value;
    return nil;
}

- (BOOL)getValueBoolAtRow:(NSInteger)row andSection:(NSInteger)section
{
    NSNumber *number = [self getValueNumberAtRow:row andSection:section];
    return [number boolValue];
}

- (void)setValue:(NSObject*)obj atRow:(NSInteger)row andSection:(NSInteger)section
{
    NSString *key = self.properties[row];
    [self.data setValue:obj forKey:key];
}

- (NSString*)displayName:(id<ITableAdapterRowConfigurator>)configurator row:(NSInteger)row section:(NSInteger)section
{
    NSString *key = self.properties[row];
    TableAdapterRowConfig*config = [configurator configForRow:key];
    if (config!=nil && config.displayName!=nil)
        return config.displayName;
    return key;
}

- (NSString*)displayDate:(id<ITableAdapterRowConfigurator>)configurator row:(NSInteger)row section:(NSInteger)section date:(NSDate*)date rowType:(TableRowType)rowType
{
    NSString *propertyName = [self rowName:row andSection:section];
    TableAdapterRowConfig *s = [self rowSetting:configurator propertyName:propertyName];
    NSString *format = nil;
    switch (rowType)
    {
        case TableRowTypeDate:
            format = @"d MMM yyyy";
            break;
        case TableRowTypeTime:
            format = @"HH:mm";
            break;
        case TableRowTypeDateTime:
            format = @"HH:mm d MMM yyyy";
            break;
        default: break;
    }
    if (s != nil)
    {
        switch (rowType)
        {
            case TableRowTypeDate:
                if (s.dateFormat!=nil)
                    format = s.dateFormat;
                break;
            case TableRowTypeTime:
                if (s.timeFormat!=nil)
                    format = s.timeFormat;
                break;
            case TableRowTypeDateTime:
                if (s.dateTimeFormat!=nil)
                    format = s.dateTimeFormat;
                break;
            default: break;
        }
    }
    return format == nil ? [date description] : [[self formatterForFormat:format] stringFromDate:date];
}

- (NSString*)rowName:(NSInteger)row andSection:(NSInteger)section
{
    NSString *propertyName = self.properties[row];
    return propertyName;
}

- (TableAdapterRowConfig*)rowSetting:(id<ITableAdapterRowConfigurator>)configurator propertyName:(NSString*)propertyName
{
    TableAdapterRowConfig *settings = nil;
    if ([self.data respondsToSelector:@selector(configForRow:)])
    {
        id<ITableAdapterRowConfigurator>temp = (id<ITableAdapterRowConfigurator>)self.data;
        settings = [temp configForRow:propertyName];
    }
    else if (configurator != nil)
        settings = [configurator configForRow:propertyName];
    return settings;
}

- (TableRowType)rowType:(id<ITableAdapterRowConfigurator>)configurator row:(NSInteger)row section:(NSInteger)section
{
    NSString *propertyName = [self rowName:row andSection:section];
    NSObject *prop = [self getValueAtRow:row andSection:section];
    TableAdapterRowConfig*s = [self rowSetting:configurator propertyName:propertyName];
    if (s != nil && s.rowType != TableRowTypeUnknown)
        return s.rowType;
    TableRowType rowType = self.defaultStringRowType;
    if ([prop isKindOfClass:[NSString class]])
        rowType = self.defaultStringRowType;
    else if ([prop isKindOfClass:[NSNumber class]])
        rowType = TableRowTypeCheckbox;
    else if ([prop isKindOfClass:[NSDate class]])
        rowType = TableRowTypeDateTime;
    return rowType;
}

- (BOOL)rowEditable:(id<ITableAdapterRowConfigurator>)configurator row:(NSInteger)row section:(NSInteger)section
{
    //NSObject *prop = nil; // TODO: Get Property
    NSString *propertyName = [self rowName:row andSection:section];
    //var editable = prop.CanWrite && prop.SetMethod.IsPublic;
    BOOL editable = YES;
    TableAdapterRowConfig *settings = [self rowSetting:configurator propertyName:propertyName];
    if (settings!=nil)
        editable = settings.editable;
    return editable;
}

@end
